src/payment/strategies/apple-pay/apple-pay-payment-strategy.ts

```diff
@@ -0,0 +1,166 @@
+                'amex',
+                'discover',
+            ],
+            lineItems: [],
+            total: {
+                label: 'Total',
+                amount: `${cart.cartAmount.toFixed(cart.currency.decimalPlaces)}`,
+                type: 'final',
+            },
+        };
+    }
+​
+    private _handleApplePayEvents(applePaySession: ApplePaySession, paymentMethod: PaymentMethod, promise: ApplePayPromise) {
+        applePaySession.onvalidatemerchant = async () => {
+            try {
+                const { body: merchantSession } = await this._onValidateMerchant(paymentMethod);
+                applePaySession.completeMerchantValidation(merchantSession);
+            } catch (err) {
+                throw new Error('Merchant validation failed');
+            }
+        };
+​
+        applePaySession.oncancel = async () =>
+            promise.reject(new PaymentMethodCancelledError('Continue with applepay'));
+​
+        applePaySession.onpaymentauthorized = (event: ApplePayJS.ApplePayPaymentAuthorizedEvent) =>
+            this._onPaymentAuthorized(event, applePaySession, paymentMethod, promise.resolve);
+    }
+​
+    private async _onValidateMerchant(paymentData: PaymentMethod) {
+        const body = [
+            `validationUrl=${appleValidationUrl}`,
+            `merchantIdentifier=${paymentData.initializationData.merchantId}`,
+            `displayName=${paymentData.initializationData.storeName}`,
+            `domainName=${window.location.hostname}`,
+        ].join('&');
+​
+        return this._requestSender.post(validationEndpoint, {
+            credentials: false,
+            headers: {
+                Accept: 'application/json',
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'X-XSRF-TOKEN': null,
+            },
+            body,
+        });
+    }
+​
+    private async _onPaymentAuthorized(
+        event: ApplePayJS.ApplePayPaymentAuthorizedEvent,
+        applePaySession: ApplePaySession,
+        paymentMethod: PaymentMethod,
+        resolve: ApplePayPromise['resolve']
+    ) {
+        const { token } = event.payment;
+        const payment: Payment = {
+            methodId: paymentMethod.id,
+            paymentData: {
+                formattedPayload: {
+                    apple_pay_token: {
+                        payment_data: token.paymentData,
+                        payment_method: token.paymentMethod,
+                        transaction_id: token.transactionIdentifier,
+                    },
+                },
+            },
+        };
+
+        return this._store.dispatch(this._paymentActionCreator.submitPayment(payment))
+            .then(() => {
+                applePaySession.completePayment(ApplePaySession.STATUS_SUCCESS);
+
+                return resolve(this._store.getState());
+            })
+            .catch(() => {
+                applePaySession.completePayment(ApplePaySession.STATUS_FAILURE);
```

Comment:
This needs to return a rejection otherwise the outer promise will resolve and trigger the success path.